experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,258 @@
+    Small = 1,
+}
+
+// ValueWidth trait
+
+// AsULE is AsUnalignedLittleEndian, i.e. "allowed in a zerovec"
+pub trait ValueWidth: Copy + zerovec::ule::AsULE {
+    const ENUM_VALUE: ValueWidthEnum;
+    fn cast_to_widest(self) -> u32;
+}
+
+impl ValueWidth for u8 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits8;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u16 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits16;
+
+    fn cast_to_widest(self) -> u32 {
+        self as u32
+    }
+}
+
+impl ValueWidth for u32 {
+    const ENUM_VALUE: ValueWidthEnum = ValueWidthEnum::Bits32;
+
+    fn cast_to_widest(self) -> u32 {
+        self
+    }
+}
+
+// TrieType trait
+
+pub trait TrieType {
+    const FAST_MAX: u32;
+    const ENUM_VALUE: TrieTypeEnum;
+}
+
+pub struct Fast;
+impl TrieType for Fast {
+    const FAST_MAX: u32 = FAST_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Fast;
+}
+
+pub struct Small;
+impl TrieType for Small {
+    const FAST_MAX: u32 = SMALL_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+// TODO: add Rust-doc that includes examples
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    /// Create a new [`CodePointTrie`] backed by borrowed data for the `index`
+    /// array and `data` array.
+    pub fn try_new(
+        header: CodePointTrieHeader,
+        index: ZeroVec<'trie, u16>,
+        data: ZeroVec<'trie, W>,
+    ) -> Result<CodePointTrie<'trie, W, T>, Error> {
+        if header.data_length < ERROR_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason: "Data array must be large enough to contain error value",
+            });
+        }
+
+        if header.data_length < HIGH_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason:
+                    "Data array must be large enough to contain value for range highStart..U+10FFFF",
+            });
+        }
+
+        if index.len() as u32 != header.index_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of index array does not match corresponding header value",
+            });
+        }
+
+        // Index array values are 16-bit width, but in one special case they
+        // are interpreted as 18-bit values. Therefore, the data array length
+        // cannot be more than 2^18.
+        if data.len() as u32 > 262144 {
```

Comment:
- For a *limit*, test with `>=`.
- Hex values are much more readable for such numbers: 0x40000
- The data array length could actually be slightly larger than this: The largest index would be 0x3ffff, but the limit (exclusive end) of the block would be higher by the block length. If I remember correctly, the data blocks at the upper end are all for supplementary code points, so the block length is 16. That gets us to a limit of 0x4000f.
- There are also some high-range and error values at the end, so the real limit needs to include those, too.
- We need not care if the data array is too long. Wasted space is harmless.
- I would start with checking no more than the C++/Java code.

Reply:
Oh, okay. Right, at the end of the data array are the highValue, errorValue, and optional padding to ensure a 32-bit alignment (is that it?). But the higher-level takeaway seems to be that, because wasted space is okay, and this check doesn't exist in C++/Java code, we can start off by removing this check as a preferable option to asserting an incorrect limit.